File: src/lib/source-config.mjs
// 仅用于识别旧版书架的来源；新配置始终留空，不以此作为联网默认值。
export const LEGACY_SOURCE_ORIGIN = 'https://www.biquge001.com';

export function normalizeSourceURL(value) {
  if (typeof value !== 'string') throw new Error('请输入网站首页的 HTTPS 地址。');
  const input = value.trim();
  if (!input) return '';
  let url;
  try { url = new URL(input); } catch { throw new Error('请输入完整的网站首页地址，例如 https://example.com/。'); }
  const labels = url.hostname.split('.');
  if (url.protocol !== 'https:' || url.username || url.password || url.port || url.search || url.hash || url.pathname !== '/'
    || url.hostname.length > 253 || labels.length < 2 || labels.some(label => !/^[a-z0-9](?:[a-z0-9-]{0,61}[a-z0-9])?$/.test(label))
    || /^[\d.]+$/.test(url.hostname) || url.hostname.endsWith('.localhost') || url.hostname.endsWith('.local')) {
    throw new Error('请填写公网网站的 HTTPS 首页地址，不含账号、端口、路径或查询参数。');
  }
  return url.origin;
}

File: backend/source_config.go
package main

import (
	"crypto/sha256"
	"errors"
	"fmt"
	"net"
	"net/url"
	"regexp"
	"strings"
)

// 仅识别旧版书籍与缓存，不作为新请求或新设置的默认网址。
const legacySourceOrigin = "https://www.biquge001.com"

var sourceLabelPattern = regexp.MustCompile(`^[a-z0-9](?:[a-z0-9-]{0,61}[a-z0-9])?$`)

func normalizeSourceOrigin(raw string) (string, error) {
	u, err := url.Parse(strings.TrimSpace(raw))
	if err != nil {
		return "", errors.New("书源网址无效，请填写 HTTPS 网站首页")
	}
	host := strings.ToLower(u.Hostname())
	if u.Scheme != "https" || u.User != nil || u.RawQuery != "" || u.ForceQuery || u.Fragment != "" || (u.Path != "" && u.Path != "/") || u.RawPath != "" || (u.Port() != "" && u.Port() != "443") || len(host) > 253 || !strings.Contains(host, ".") || net.ParseIP(host) != nil || strings.HasSuffix(host, ".localhost") || strings.HasSuffix(host, ".local") {
		return "", errors.New("请填写公网网站的 HTTPS 首页地址，不含账号、端口、路径或查询参数")
	}
	for _, label := range strings.Split(host, ".") {
		if !sourceLabelPattern.MatchString(label) {
			return "", errors.New("网站域名无效")
		}
	}
	return "https://" + host, nil
}

func bookOrigin(b book) string {
	if b.SourceOrigin == "" {
		return legacySourceOrigin
	}
	return b.SourceOrigin
}
func sourceCacheKey(origin, path string) string {
	if origin == legacySourceOrigin {
		return onlineID(path)
	}
	return fmt.Sprintf("%x-%s", sha256.Sum256([]byte(origin)), onlineID(path))
}
